src/platform/plugins/shared/dashboard/common/lib/dashboard_section_converters.ts

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import { DashboardSectionMap, DashboardSectionState } from '../dashboard_container/types';
+
+export const convertSectionArrayToSectionMap = (
```

Comment:
Why are we storing sections as an array and then converting back to map on read? I know this is how panels is done, but is that best? Would it be possible to just store sections as a map and avoid these conversions?

Reply:
This was done on recommendation from @nickpeihl for the Dashboard API.

Reply:
When a user is composing a request body to create/update a dashboard using the public HTTP API, it is preferable to use arrays rather than maps because we can make the `sections[].id` optional for the consumer and generate a UUID if not provided, (e.g. `sections: [ { ...state, id: id ?? uuidv4 } ]`. With a map the consumer is required to set a unique key (e.g. `sections: { "unique-001": { <section-state> }`) when composing the request body.

@Heenawter Is it necessary to convert the array to a map for the UI state? I know we do it for panels, so I guess consistency maybe?

Reply:
I originally had it as an array, but there are a few reasons I decided to convert it to a map in runtime state:
1. Consistency, like you said
2. We have to look up sections based on their ID in order to determine if they are collapsed or not (for example, in `untilEmbeddableLoaded`, where we ignore panels in collapsed sections), and it is easier + faster to do this with a map rather than searching through an array every time for the right ID
3. Arrays in runtime state have been known to cause comparison issues in the past, since they care about order; but if the contents of the array are the same, we don't typically care if the items are at different indices. Maps are safer for this